File: basic-algorithms/programming-assignment-6/Solution.java
import java.io.*;
import java.util.*;

public class Solution {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);

        // Read input
        int degree = input.nextInt();
        long[] f = new long[degree + 1];
        long[] g = new long[degree + 1];

        for (int i = 0; i < degree + 1; i++)
            f[i] = input.nextLong();
        
        for (int i = 0; i < degree + 1; i++)
            g[i] = input.nextLong();

        input.close();
        
        // Multiply
        long[] product = karatsubaMultiplication(f, g);

        for (int i = 0; i < product.length; i++) {
            System.out.print(product[i]);
            if (i != product.length - 1)
                System.out.print(' ');
        }
    }

    static long[] naiveMultiplication(long[] f, long[] g) {
        // Initialize product
        long[] product = new long[2*f.length - 1];
        for (int i = 0; i < product.length; i++)
            product[i] = 0;

        // Multiply
        for (int i = 0; i < g.length; i++)
            for (int j = 0; j < f.length; j++)
                product[i + j] = product[i + j] + g[i]*f[j];

        return product;
    }

    static long[] karatsubaMultiplication(long[] f, long[] g) {
        // Recursively call karatsubaMultiplication if degree is greater than 10. Else, call naiveMultiplication
        int degree = f.length - 1;
        if (degree > 64) {
            // Split polynomials
            long[] fLo = Arrays.copyOfRange(f, 0, degree/2 + 1);
            long[] fHi = Arrays.copyOfRange(f, degree/2 + 1, f.length);

            long[] gLo = Arrays.copyOfRange(g, 0, degree/2 + 1);
            long[] gHi = Arrays.copyOfRange(g, degree/2 + 1, g.length);

            // Compute sums
            long[] F = new long[fHi.length];
            long[] G = new long[gHi.length];
            for (int i = 0; i < F.length; i++) {
                F[i] = fHi[i] + fLo[i];
                G[i] = gHi[i] + gLo[i];
            } 

            // Multiply
            long[] U = karatsubaMultiplication(fHi, gHi);
            long[] V = karatsubaMultiplication(fLo, gLo);
            long[] W = karatsubaMultiplication(F, G);

            long[] product = new long[2*degree + 1];
            for (int i = 0; i < product.length; i++)
                product[i] = 0;

            for (int i = 0; i < V.length; i++)
                product[i] = V[i];
            
            for (int i = 0; i < U.length; i++)
                product[i + degree/2 + 1] += W[i] - U[i] - V[i];
            
            for (int i = 0; i < U.length; i++)
                product[i + degree + 1] += U[i];

            return product;
        }
        else {
            // Multiply
            return naiveMultiplication(f, g);
        }
    }
}

File: basic-algorithms/programming-assignment-6/solution.py
# import dependencies
import sys
import copy

def naive_multiplication(f, g):
    # initialize product
    product = [0]*(2*len(f) - 1)

    # multiply
    for i in range(len(g)):
        for j in range(len(f)):
            product[i + j] = product[i + j] + g[i]*f[j]

    return product

def karatsuba_multiplication(f, g):
    # use recursion for large degrees and naive_multiplication for smaller degrees
    degree = len(f) - 1
    if degree > 64:
        # split polynomials
        f_lo = f[:int(degree/2) + 1]
        f_hi = f[int(degree/2) + 1:]

        g_lo = g[:int(degree/2) + 1]
        g_hi = g[int(degree/2) + 1:]
    
        # compute sums
        F = copy.deepcopy(f_hi)
        G = copy.deepcopy(g_hi)
        
        for i in range(len(f_lo)):
            F[i] += f_lo[i]
            G[i] += g_lo[i]

        # multiply
        U = karatsuba_multiplication(f_hi, g_hi)
        V = karatsuba_multiplication(f_lo, g_lo)
        W = karatsuba_multiplication(F, G)

        product = [0]*(2*degree + 1)
        for i in range(len(V)):
            product[i] = V[i]
        
        for i in range(len(U)):
            product[i + int(degree/2) + 1] += W[i] - U[i] - V[i]

        for i in range(len(U)):
            product[i + degree + 1] += U[i]

        return product
    else:
        # multiply
        return naive_multiplication(f, g)


def main():
    # read input
    input_data = sys.stdin.read().split('\n')

    degree = int(input_data[0])

    f = []
    for coefficient in input_data[1].split(' '):
        f.append(int(coefficient))

    g = []
    for coefficient in input_data[2].split(' '):
        g.append(int(coefficient))

    # multiply
    product = karatsuba_multiplication(f, g)

    # print output
    sys.stdout.write(' '.join(str(coefficient) for coefficient in product))

main()